libshaderc/src/shaderc.cc

```diff
@@ -159,42 +160,81 @@ class StageDeducer {
+           result_releaser_ != nullptr;
   }
 
-  // Find filename in search path and returns its contents.
-  std::pair<std::string, std::string> include_delegate(
-      const char* filename) const override {
-    if (!AreValidCallbacks())
-      return std::make_pair<std::string, std::string>(
-          "", "unexpected include directive");
-    shaderc_includer_response* data =
-        get_includer_response_(user_data_, filename);
-    std::pair<std::string, std::string> entry =
-        std::make_pair(std::string(data->path, data->path_length),
-                       std::string(data->content, data->content_length));
-    release_includer_response_(user_data_, data);
-    return entry;
+  // Maps a shaderc_include_type to the correpsonding Glslang include type.
+  shaderc_include_type GetIncludeType(
+      glslang::TShader::Includer::IncludeType type) {
+    switch (type) {
+      case glslang::TShader::Includer::EIncludeRelative:
+        return shaderc_include_type_relative;
+      case glslang::TShader::Includer::EIncludeStandard:
+        return shaderc_include_type_standard;
+      default:
+        break;
+    }
+    assert(0 && "Unhandled shaderc_include_type");
+    return shaderc_include_type_relative;
+  }
+
+  // Resolves an include request for the requested source of the given
+  // type in the context of the specified requesting source.  On success,
+  // returns a newly allocated IncludeResponse containing the fully resolved
+  // name of the requested source and the contents of that source.
+  // On failure, returns a newly allocated IncludeResponse where the
+  // resolved name member is an empty string, and the contents members
+  // contains error details.
+  virtual glslang::TShader::Includer::IncludeResult* include_delegate(
+      const char* requested_source,
+      glslang::TShader::Includer::IncludeType type,
+      const char* requesting_source,
+      size_t include_depth) override {
+    if (!AreValidCallbacks()) {
+      const char kUnexpectedIncludeError[] =
+          "#error unexpected include directive";
+      return new glslang::TShader::Includer::IncludeResult{
+          "", kUnexpectedIncludeError, strlen(kUnexpectedIncludeError),
+          nullptr};
+    }
+    shaderc_include_result* include_result =
+        resolver_(user_data_, requested_source, GetIncludeType(type),
+                  requesting_source, include_depth);
+    // Make a glslang IncludeResult from a shaderc_include_result.  The
+    // user_data member of the IncludeResult is a pointer to the
+    // shaderc_include_result object, so we can later release the latter.
+    return new glslang::TShader::Includer::IncludeResult{
+        std::string(include_result->source_name,
+                    include_result->source_name_length),
+        include_result->content, include_result->content_length,
+        include_result};
+  }
+
+  // Releases the given IncludeResult.
+  virtual void release_delegate(
```

Comment:
But do we need to take care of the new-ed pointer of glslang::TShader::Includer::IncludeResult? If release_delegate() is missed, will the space created in line:219 be destroyed?